fix: (Core) bug fixes for fixed card layout

DeepakSap14 opened the pull request:
#### Please provide a link to the associated issue.
#3603  #3600  #3594  #3595 #3552 #3542 

#### Please provide a brief summary of this pull request.
This PR addresses fix for bugs raised against fixed card layout component.
Above mentioned issues has been fixed in this PR.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] update `README.md`
- [x] Documentation Examples
- [x] Stackblitz works for all examples

KevinOkamoto reviewed (commented):
Drag-n-drop does not seem to be working.

DeepakSap14 commented:
> Drag-n-drop does not seem to be working.

For scrolling to work in touch devices. Only option we have to delay the drag. so currently in examples default dragStartDelay input property is set to 500ms. If you keep the mouse on card for that long then card drag starts.

KevinOkamoto commented:
> > Drag-n-drop does not seem to be working.
> 
> For scrolling to work in touch devices. Only option we have to delay the drag. so currently in examples default dragStartDelay input property is set to 500ms. If you keep the mouse on card for that long then card drag starts.

Hmm ... can we shorten the delay to 100ms?

DeepakSap14 commented:
> > > Drag-n-drop does not seem to be working.
> > 
> > 
> > For scrolling to work in touch devices. Only option we have to delay the drag. so currently in examples default dragStartDelay input property is set to 500ms. If you keep the mouse on card for that long then card drag starts.
> 
> Hmm ... can we shorten the delay to 100ms?

scroll did not work on 100ms. I have made it 200ms default value. Scroll and drag both are working fine now. It can be set by user input as well.

DeepakSap14 commented:
Issue regarding space between items is fixed now.

KevinOkamoto reviewed (approved):
Looks good.

InnaAtanasova reviewed (changes requested):
Some issue I found testing on mobile device:
- content is overflowing
![AAAC60E6-B508-41B1-89E2-F762A14DA935](https://user-images.githubusercontent.com/39598672/96755120-4b108100-13a0-11eb-924c-5499f0650d07.jpg)

- in landscape mode the example doesn't look good:
![792D67D6-E5CF-4A17-815B-6F0FB6BF782B](https://user-images.githubusercontent.com/39598672/96755194-654a5f00-13a0-11eb-90e9-9bee16763cbb.jpg)
![2325FBCB-7207-4A23-B9C0-5963C7EAD341](https://user-images.githubusercontent.com/39598672/96755249-75fad500-13a0-11eb-8b9a-0fd767cb867d.jpg)

- The issue created by @dimamarksman is not really fixed https://github.com/SAP/fundamental-ngx/issues/3600
when I try to scroll the content with my finger it either does not respond to the event ir starts dragging the card.

- Desktop: (the container should shrink when the docs site is resized)
<img width="761" alt="Screen Shot 2020-10-21 at 1 07 25 PM" src="https://user-images.githubusercontent.com/39598672/96755943-5b752b80-13a1-11eb-858b-6848c34ac10b.png">

DeepakSap14 commented:
@InnaAtanasova, In example created for Mobile i have fixed width and height. so

1. content is overflowing
- - Mobile example is just mimicking mobile height and width. So this example has fixed width, manually set to const portraitWidth = 375; and const landscapeWidth = 812. example content is considering this width and height. Thus in webpage it looks overflow. 
Usually for verifying mobile, webpage should be opened in mobile devices. But testing team member insisted on having Mobile example here.

2. in landscape mode the example doesn't look good:
-- same as point 1. It can be verified with another examples present, by opening in Mobile devices.

3. issue is not really fixed #3600 
-- Scroll is working for me in mobile device. Now i have made dragStartDelay="300ms" for the examples created. it should scroll now and also on long press drag will start. (Mobile touch sensitivity may be issue here)

4. Desktop: (the container should shrink when the docs site is resized)
-- same as point 1. Because of fixed width, it is not shrinking. Please verify with another examples present, by opening in Mobile devices.

InnaAtanasova commented:
I am well aware that the container has fixed sizes and that's why it overflows :) 
Since you are mentioning other examples in mobile mode I can point you to Menu in Mobile mode (https://sap.github.io/fundamental-ngx/#/core/menu#mobile-menu), Select in Mobile view (https://sap.github.io/fundamental-ngx/#/core/select#select-mobile), Dialog in Mobile mode (https://sap.github.io/fundamental-ngx/#/core/dialog#dialogMobile) and many others where the user is given an option to preview and test the component in this mode. This is Fundamental-ngx, not Fundamental-styles where we avoid adding logic in the documentation. Please verify again the examples. Thanks

DeepakSap14 commented:
> I am well aware that the container has fixed sizes and that's why it overflows :)
> Since you are mentioning other examples in mobile mode I can point you to Menu in Mobile mode (https://sap.github.io/fundamental-ngx/#/core/menu#mobile-menu), Select in Mobile view (https://sap.github.io/fundamental-ngx/#/core/select#select-mobile), Dialog in Mobile mode (https://sap.github.io/fundamental-ngx/#/core/dialog#dialogMobile) and many others where the user is given an option to preview and test the component in this mode. This is Fundamental-ngx, not Fundamental-styles where we avoid adding logic in the documentation. Please verify again the examples. Thanks

Thanks! I will do the same way.

DeepakSap14 commented:
> > 1. issue is not really fixed #3600
> >    -- Scroll is working for me in mobile device. Now i have made dragStartDelay="300ms" for the examples created. it should scroll now and also on long press drag will start. (Mobile touch sensitivity may be issue here)
> 
> As for the #3600 I am still having issues with the scroll on both iPhone 11 Pro max and iPhone 12.
> When you open the page first it works ok, you are able to scroll. When you drag and drop a card in any of the examples it becomes almost impossible (or with a big delay), to scroll unless you touch the area outside of the container, where the documentation scroll is. If we have problems with these 2 devices we have a big problem (in my opinion).

Okay. will check it.

salarenko reviewed (changes requested):
- For the future, please name branches properly:
`(fix | feat | chore)/(ISSUE_NUMBER)-(PR_NAME)`

- Its a bit subjective, but I think we should have drag delay only for mobile devices

- I don't know the exact spec, but I've checked the examples of card grid and for every case the grid and the cards seems to take all of the available space.
In current implementation:
```
<fd-fixed-card-layout>    (display: flex)
    <card-column>         (collapses to minimum width of the content)
        <fd-card>         (defines the minimum width)
```
So we can say that the cards are displayed almost as in fixed size. We could change this behavior by making the card-columns grow horizontally (`flex-grow: 1`) to stretch the layout grid and fill all of the available content. Maybe this is the desired behavior?

Current implementation:
![image](https://user-images.githubusercontent.com/17496353/96839743-e1e54800-1449-11eb-8df6-607f059e3ef9.png)

With growing columns:
![image](https://user-images.githubusercontent.com/17496353/96839818-f9243580-1449-11eb-8b75-08de6fef8b68.png)

- I see you've created some examples with `angular-resize-event`. This works, but I think our grid should have this feature as part of its implementation. Maybe we could use [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) here? This is a bit problematic case and we should not bring any third party libraries into the library or the documentation without discussing the problem first. I will bring this issue in today's meeting.

DeepakSap14 commented:
@mich

> > 1. issue is not really fixed #3600
> >    -- Scroll is working for me in mobile device. Now i have made dragStartDelay="300ms" for the examples created. it should scroll now and also on long press drag will start. (Mobile touch sensitivity may be issue here)
> 
> As for the #3600 I am still having issues with the scroll on both iPhone 11 Pro max and iPhone 12.
> When you open the page first it works ok, you are able to scroll. When you drag and drop a card in any of the examples it becomes almost impossible (or with a big delay), to scroll unless you touch the area outside of the container, where the documentation scroll is. If we have problems with these 2 devices we have a big problem (in my opinion).

@InnaAtanasova 
I have create mobile display example now. checked in 2 mobile phones for scroll and drag, it works fine. please let me know if you still see the issue.

DeepakSap14 commented:
> * For the future, please name branches properly:
>   `(fix | feat | chore)/(ISSUE_NUMBER)-(PR_NAME)`
> * Its a bit subjective, but I think we should have drag delay only for mobile devices
> * I don't know the exact spec, but I've checked the examples of card grid and for every case the grid and the cards seems to take all of the available space.
>   In current implementation:
> 
> ```
> <fd-fixed-card-layout>    (display: flex)
>     <card-column>         (collapses to minimum width of the content)
>         <fd-card>         (defines the minimum width)
> ```
> 
> So we can say that the cards are displayed almost as in fixed size. We could change this behavior by making the card-columns grow horizontally (`flex-grow: 1`) to stretch the layout grid and fill all of the available content. Maybe this is the desired behavior?
> 
> Current implementation:
> ![image](https://user-images.githubusercontent.com/17496353/96839743-e1e54800-1449-11eb-8df6-607f059e3ef9.png)
> 
> With growing columns:
> ![image](https://user-images.githubusercontent.com/17496353/96839818-f9243580-1449-11eb-8b75-08de6fef8b68.png)
> 
> * I see you've created some examples with `angular-resize-event`. This works, but I think our grid should have this feature as part of its implementation. Maybe we could use [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) here? This is a bit problematic case and we should not bring any third party libraries into the library or the documentation without discussing the problem first. I will bring this issue in today's meeting.

@salarenko 
1. Noted.
2. dragStartDelay is an input property. Application developer will be able to set if it is desktop or mobile accordingly. let me know if you want some change here.
3. Earlier i was using flex 1. which distributes all available spaces among flex-items. but with that mentioned issue arises https://github.com/SAP/fundamental-ngx/issues/3595. Because cards has max-width defined, Some extra space is taken by flex-items. by making flex 0, i am making card fixed at min-width, which is a cost for avoiding issue #3595. I will raise a issue and drop a mail to designers on the limitations.
4.  "angular-resize-event" : I have updated the **Notes** on the documentation about this library usages after Discussion on component review meeting.
5. removed custom dragStartDelay.

DeepakSap14 commented on apps/docs/src/app/core/component-docs/fixed-card-layout/examples/default/fixed-card-layout-examples.component.html:
```diff
@@ -10,11 +10,149 @@
-    <ng-container *ngIf="card1Visibility">
-        <fd-card *fdCardDef>
+<div (resized)="onResized()" style="padding: 1rem; overflow: auto;">
+    <fd-fixed-card-layout [dragStartDelay]="300">
```
removed custom dragStartDelay.
